querydsl-sql/src/main/java/com/querydsl/sql/types/AbstractDateTimeType.java

```diff
@@ -38,14 +37,40 @@ protected static Calendar utc() {
         return (Calendar) UTC.clone();
     }
 
-    protected static final DateTimeFormatter dateFormatter = DateTimeFormat.forPattern("yyyy-MM-dd");
+    private static final ThreadLocal<SimpleDateFormat> dateFormatter = new ThreadLocal<SimpleDateFormat>() {
```

Comment:
We're setting the minimal JRE requirement to 8 in the upcoming QueryDSL release. As a result, I played with swapping out the `SimpleDateFormat` here with Java 8's `DateTimeFormatter`. However, that only works with `Temporals` and not simple `Date` objects. Converting a `Date` to a `Temporal` seems rather involved.

@daniel-shuy WDYT? Should we swap the implementation out for  `DateTimeFormatter`?

Reply:
> We're setting the minimal JRE requirement to 8 in the upcoming QueryDSL release

Nice! I've rebased this PR onto the `java-8-2` branch.

> As a result, I played with swapping out the SimpleDateFormat here with Java 8's DateTimeFormatter. However, that only works with Temporals and not simple Date objects. Converting a Date to a Temporal seems rather involved.

`java.util.Date` can be easily converted to `java.time.Instant` using [Date#toInstant()](https://docs.oracle.com/javase/8/docs/api/java/util/Date.html#toInstant--). `Calendar` also has an equivalent [Calendar#toInstant()](https://docs.oracle.com/javase/8/docs/api/java/util/Calendar.html#toInstant--) method.

`java.sql.Timestamp`, `java.sql.Date` and `java.sql.Time` also have [Timestamp#toLocalDateTime()](https://docs.oracle.com/javase/8/docs/api/java/sql/Timestamp.html#toLocalDateTime--), [Date#toLocalDate()](https://docs.oracle.com/javase/8/docs/api/java/sql/Date.html#toLocalDate--) and [Time#toLocalTime()](https://docs.oracle.com/javase/8/docs/api/java/sql/Time.html#toLocalTime--) to convert to `LocalDateTime`, `LocalDate` and `LocalTime` respectively.

> WDYT? Should we swap the implementation out for DateTimeFormatter?

I've added a commit that replaces `SimpleDateFormat` with Java 8's `DateTimeFormatter`, you can take a look. If we decide not to use it I can simply drop the commit. One of the advantages is that it allows [AbstractJSR310DateTimeType](https://github.com/querydsl/querydsl/pull/2643/commits/e19073171abfcf4caaf6894eb7eb652a53f04b8b#diff-4cc25c794652fea9373b3da85194ed1817806a7637a43a1f9d76d8e96af9dccc) to extend [AbstractDateTimeType](https://github.com/querydsl/querydsl/pull/2643/commits/e19073171abfcf4caaf6894eb7eb652a53f04b8b#diff-3237b2c366f99970fac048ab8af05274f5e12b4377789411a8d0d1aceaf3dac9).

Reply:
Thanks! I agree we should include these changes. I also like how `AbstractJSR310DateTimeType` can now extend `AbstractDateTimeType`.

I see the tests is failing. The CUBRID tests regularly fail (Ugh), so you don't need to check those. The other ones I still have to look into. Unless you're already working on them?

Reply:
Sorry, I've fixed it. Turns out `Instant` has to be converted to `LocalDateTime` first for it to be formatted using `DateTimeFormatter`.

Some tests are still failing, but I think they are existing issues in the `java-8-2` branch.

Reply:
Thanks! Ill look into these and merge the PR tomorrow.